Introduction of STM32WB0 series

mathieuchopstm opened the pull request:
This PR contains the basic groundwork needed to support the STM32WB0 series in Zephyr:
* Bindings for WB09-specific RCC / GPIO INTC / flash controller
* Base DTSI for all SoCs in the series
* SoC support code
* Clock control driver
* GPIO interrupt controller driver
* Flash controller driver
* Compatibility patches for GPIO driver
* Support for two Nucleo-64 boards
  * Nucleo-WB05KZ
  * Nucleo-WB09KE

This PR does not need to bring any modification to the STM32 UART driver, as the IP in STM32WB0 is compatible with that of other series.

mathieuchopstm committed:
west.yml: hal_stm32: Update to HAL integrating STM32WB0

Update the West manifest to point to an updated version of the STM32 HAL
compatible with WB0 series thank to inclusion of STM32CubeWB0 package.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
soc: st: stm32: add STM32WB0 series

Adds support for the STM32WB0 MCU series.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
dts: bindings: clock: add STM32WB0 RCC and LSI

Add the Device Tree bindings for the RCC and LSI clock of STM32WB0 series.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
drivers: clock: add STM32WB0 clock control

Add control driver for STM32WB0 series, with support for all clock sources.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
include: dt-bindings: add STM32WB0 reset header

Adds the reset controller dt-bindings header for STM32WB0 series.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
dts: arm: st: wb0: add DTSI for STM32WB0 series

Adds Device Tree include files for all MCUs in the STM32WB0 series.
These DTSI files only contain the supported peripherals for now.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
boards: st: nucleo_wb09ke: add support for Nucleo-WB09KE board

This commit adds support for the ST Nucleo-WB09KE board.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
drivers: flash: stm32: add STM32WB0 flash controller

Adds a basic driver for the STM32WB0 flash controller (read/erase/write).
Extended operations are not supported by this driver.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
dts: bindings: flash: add STM32WB0 flash controller

Add the Device Tree binding for the STM32WB0 flash controller.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
modules: Kconfig.stm32: Add Kconfig symbols for RADIO

The RADIO and RADIO_TIMER HAL modules will be required to
implement BLE support for STM32WB0 series.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
drivers: intc: add STM32WB0 GPIO interrupt controller

Adds a driver for the STM32WB0 series GPIO interrupt controller.
This driver implements the STM32 GPIO INTC API, along with an extension
function used to check if a specific line is available on current board.

This also extends the GPIO INTC API to support level-sensitive interrupts,
as this feature is available on STM32WB0.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
drivers: gpio: stm32: add support for STM32WB0

Adds support for the STM32WB0 series in the existing STM32 GPIO driver.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
dts: bindings: intc: add STM32WB0 GPIO interrupt controller

Add the Device Tree binding for the STM32WB0 GPIO interrupt controller.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
boards: st: nucleo_wb05kz: add support for Nucleo-WB05KZ board

Add support for the STMicroelectronics Nucleo-WB05KZ board.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

erwango commented:
See https://github.com/zephyrproject-rtos/hal_stm32/pull/218 for the PAGESIZE issue

erwango commented on dts/arm/st/wb0/stm32wb09.dtsi:
```diff
@@ -0,0 +1,27 @@
+
+#include <st/wb0/stm32wb05.dtsi>
+
+/**
+ * STM32WB09 is mostly identical to STM32WB05.
+ * The only differences are RAM size, flash size and TRNG IP.
+ */
+/ {
+	soc {
+		compatible = "st,stm32wb09", "st,stm32wb0", "simple-bus";
+
+		flash: flash-controller@40001000 {
+			flash0: flash@10040000 {
+				reg = <0x10040000 DT_SIZE_K(512)>;
+			};
+		};
+	};
+
+	sram0: memory@20000000 {
+		reg = <0x20000000 DT_SIZE_K(64)>;
+	};
+};
```
Nitpick:
Move memory defnitions in `dts/arm/st/wb0/stm32wb09Xe.dtsi` file.
No consequence and eases automated file parsing if need arises

Apply uniformously

mathieuchopstm replied:
> No consequence

It does lead to duplication as both `stm32wb06Xc` and `stm32wb07Xc` will now define flash size. Other than that, I agree.

Change will be done.

erwango commented on dts/arm/st/wb0/stm32wb0.dtsi:
```diff
@@ -0,0 +1,207 @@
+#include <freq.h>
+#include <mem.h>
+
+/* Helper macro aligned with STM32_RESET() */
+#define STM32_CLOCK(bus, bit) STM32_CLOCK_BUS_ ##bus (1 << bit)
```
Quite nice, but I'd rather like to see being applied uniformly over whole family. So, let's log an issue to be sure it doesn't get missed.

mathieuchopstm replied:
Created as #78106

erwango commented on boards/st/nucleo_wb05kz/nucleo_wb05kz_defconfig:
```diff
@@ -0,0 +1,21 @@
+# Enable GPIO
+CONFIG_GPIO=y
+
+# Enable pin controller
+CONFIG_PINCTRL=y
```
Could be removed

erwango commented on boards/st/nucleo_wb05kz/nucleo_wb05kz.yaml:
```diff
@@ -0,0 +1,13 @@
+ram: 24
+flash: 192
+supported:
+  - gpio
```
also add `arduino_gpio`

erwango reviewed (changes requested):
Great work!
I skipped over the flash driver and boards for this first iteration.

mathieuchopstm commented on soc/st/stm32/stm32wb0x/Kconfig:
```diff
@@ -0,0 +1,143 @@
+config SMPS_OUTPUT_VOLTAGE_1_35V
+	bool "1.35V"
+
+config SMPS_OUTPUT_VOLTAGE_1_40V
+	bool "1.40V"
+
+config SMPS_OUTPUT_VOLTAGE_1_45V
+	bool "1.45V"
+
+config SMPS_OUTPUT_VOLTAGE_1_50V
+	bool "1.50V"
+
+config SMPS_OUTPUT_VOLTAGE_1_55V
+	bool "1.55V"
+
+config SMPS_OUTPUT_VOLTAGE_1_60V
+	bool "1.60V"
+
+config SMPS_OUTPUT_VOLTAGE_1_65V
+	bool "1.65V"
+
+config SMPS_OUTPUT_VOLTAGE_1_70V
+	bool "1.70V"
+
+config SMPS_OUTPUT_VOLTAGE_1_75V
+	bool "1.75V"
+
+config SMPS_OUTPUT_VOLTAGE_1_80V
+	bool "1.80V"
+
+config SMPS_OUTPUT_VOLTAGE_1_85V
+	bool "1.85V"
+
+config SMPS_OUTPUT_VOLTAGE_1_90V
+	bool "1.90V"
+
+config SMPS_OUTPUT_VOLTAGE_1_95V
+	bool "1.95V"
+
+endchoice
```
Possible, but would require introduction of a PWRC node.

Is that OK?

erwango replied:
Yeah, I think we'll need it sooner or later and here it seems to be good time. Maybe make it wb0 specific for now as we know alredy it diverges form other series.

mathieuchopstm commented on soc/st/stm32/Kconfig:
```diff
@@ -56,17 +56,42 @@ config STM32_WKUP_PINS
 
 config POWER_SUPPLY_DIRECT_SMPS
 	bool "Direct SMPS supply"
+	depends on !SOC_SERIES_STM32WB0X
```
i.e. `depends on (all series except WB0)`?

erwango replied:
Yes. All series using SMPS

mathieuchopstm commented on soc/st/stm32/stm32wb0x/Kconfig.defconfig:
```diff
@@ -0,0 +1,15 @@
+
+if SOC_SERIES_STM32WB0X
+
+# STM32WB0 MR_BLE radio supports only 128 simultaneous active connections
+config BT_MAX_CONN
+	range 1 128
```
Removing.

mathieuchopstm commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -550,26 +593,40 @@ static int gpio_stm32_pin_interrupt_configure(const struct device *dev,
-	default:
-		err = -EINVAL;
-		goto exit;
+	if (mode == GPIO_INT_MODE_LEVEL) {
```
Not needed due to the check above:
https://github.com/zephyrproject-rtos/zephyr/blob/8d0a91d7a399648207c47e5232d2bab354a39fff/drivers/gpio/gpio_stm32.c#L586-L589

Or would you prefer to remove that check entirely and add:
```suggestion
	if (mode == GPIO_INT_MODE_LEVEL) {
		if (!IS_ENABLED(/* ... */) {
			err = -ENOTSUP;
			goto exit;
		}
```

mathieuchopstm replied:
Reworked as proposed.

mathieuchopstm commented on soc/st/stm32/Kconfig:
```diff
@@ -56,17 +56,42 @@ config STM32_WKUP_PINS
 choice POWER_SUPPLY_CHOICE
 	prompt "STM32 power supply configuration"
-	default POWER_SUPPLY_LDO
+	default POWER_SUPPLY_LDO if !SOC_SERIES_STM32WB0X
+	default POWER_SUPPLY_SMPS_SUPPLIES_LDO if SOC_SERIES_STM32WB0X
```
Moving to `soc/st/stm32/stm32wb0x/Kconfig` appears to work. Will do the necessary changes.

However, since other power supply configuration is being moved to DTS, should I take the opportunity to also move this to DTS? (for WB0 only)

erwango replied:
If not too much trouble, why not. But not mandatory if this extends the work too much.

mathieuchopstm commented on soc/st/stm32/Kconfig:
```diff
@@ -56,17 +56,42 @@ config STM32_WKUP_PINS
 	help
 	  Enable support for STM32 PWR wake-up pins.
 
+# On STM32WB0, the default selection is 'SMPS enabled'
+# as it corresponds to reset state and works regardless of
+# hardware configuration, whereas disabling SMPS on boards
+# using the SMPS hardware configuration will "brick".
```
Will move to WB0 series Kconfig file.

mathieuchopstm committed:
dts: bindings: power: add STM32WB0 power controller

Add a Device Tree binding for the STM32WB0 power controller.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm commented on include/zephyr/dt-bindings/clock/stm32wb0_clock.h:
```diff
@@ -0,0 +1,81 @@
+/* `mask` is only 0x1 for WB06/WB07, but a single definition with mask=0x3 is acceptable */
+#define SPI3_I2S3_SEL(val)	STM32_DEVICE_ALT_CLOCK(val, 3, 22, CFGR_REG)
+
+#define MR_BLE_WB09_SEL(val)	STM32_DEVICE_ALT_CLOCK(val, 1, 2, APB2ENR_REG)
+#define MR_BLE_WB07_SEL(val)	STM32_DEVICE_ALT_CLOCK(val, 3, 1, APB2ENR_REG)
+#define MR_BLE_WB05_SEL(val)	MR_BLE_WB09_SEL(val)
+#define MR_BLE_WB06_SEL(val)	MR_BLE_WB07_SEL(val)
```
Dropped `MR_BLE_SEL` for now, postponed to introduction of BLE support.

mathieuchopstm commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -156,6 +156,49 @@ static inline uint32_t stm32_pinval_get(gpio_pin_t pin)
 
+static inline void ll_gpio_set_pin_pull(GPIO_TypeDef *GPIOx, uint32_t Pin, uint32_t Pull)
+{
+#if defined(CONFIG_GPIO_INTC_STM32WB0)
```
Changed check to `#if defined(CONFIG_SOC_SERIES_STM32WB0X)`.
Adding a binding judged overkill as WB0 should be the only series to have this particularity.

mathieuchopstm committed:
drivers: hwinfo: stm32: mark STM32WB0 series as incompatible

The existing hwinfo driver for STM32 is incompatible with STM32WB0 series.
Prevent compiling the driver if the target's series is STM32WB0.
This fixes the build failure on the drivers.hwinfo.api test.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm commented:
Hi @JarmouniA,
thanks for the reminder - I have made the changes you requested to align the boards with others.

Tested flashing with latest STM32CubeCLT, and everything works OK.

Also added the missing picture for Nucleo-WB05KZ while at it, and removed PyOCD as I don't think it supports WB0 yet.

mathieuchopstm commented:
Compressed WB05KZ picture to pass CI...

erwango commented on dts/bindings/power/st,stm32wb0-pwr.yaml:
```diff
@@ -0,0 +1,130 @@
+      This property is not valid if `smps-mode` if "OFF",
+      and required otherwise.
+    enum:
+      - 1 # 1.5µH inductance, 2.2µF output capacitance
+      - 2 # 2.2µH inductance, 4.7µF output capacitance
+      - 3 #  10µH inductance, 4.7µF output capacitance
```
Comments are not rendered in binding generated documentation. Please put these in description section.

erwango commented on drivers/flash/flash_stm32wb0x.c:
```diff
@@ -0,0 +1,475 @@
+#ifdef CONFIG_FLASH_PAGE_LAYOUT
+	.page_layout = flash_wb0x_pages_layout,
+#endif
+#ifdef CONFIG_FLASH_EX_OP_ENABLED
+	.ex_op = flash_wb0x_ex_op,
+#endif
```
API already returns `-ENOTSUP` if not defined

mathieuchopstm replied:
Good catch. Removing this useless stub.

erwango commented on boards/st/nucleo_wb09ke/nucleo_wb09ke_defconfig:
```diff
@@ -0,0 +1,14 @@
+# SPDX-License-Identifier: Apache-2.0
+
+# Enable clock
+CONFIG_CLOCK_CONTROL=y
```
Could now be removed

mathieuchopstm replied:
Removed from both boards.

erwango reviewed (changes requested):
Otherwise LGTM

mathieuchopstm commented on drivers/flash/flash_stm32wb0x.c:
```diff
@@ -0,0 +1,475 @@
+/**
+ * Driver private definitions & assertions
+ */
+#if defined(CONFIG_SOC_STM32WB09XX)
+#	define PAGES_IN_FLASH 256
+#elif defined(CONFIG_SOC_STM32WB07XX) \
+	|| defined(CONFIG_SOC_STM32WB06XX)
+#	define PAGES_IN_FLASH 128
+#elif defined(CONFIG_SOC_STM32WB05XX)
+#	define PAGES_IN_FLASH 96
```
Originally, I did not find any definition that provided the flash size as a define.
However, it appears that `_MEMORY_FLASH_SIZE_` in `system_stm32wb0x.h` could fit the bill.

Will make the change.

mathieuchopstm commented on drivers/flash/flash_stm32wb0x.c:
```diff
@@ -0,0 +1,475 @@
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/flash.h>
+#include <zephyr/drivers/flash/stm32_flash_api_extensions.h>
```
No, it's an artifact from first iteration of driver in which EXOP was planned.
Removing.

mathieuchopstm commented on dts/bindings/power/st,stm32wb0-pwr.yaml:
```diff
@@ -0,0 +1,130 @@
+
+      This property is only valid if `smps-mode` is "RUN".
+
+      If unspecified, the reset value of 1.40V is used.
```
After discussion, checks are found to be overkill - removing and switching to `default` usage.

erwango commented:
@nashif,, another question on assignment
```
Area matches: {<Area STM32 Platforms>: 49, <Area Drivers: Clock control>: 3, <Area Interrupt Handling>: 2, <Area Drivers: Flash>: 1, <Area Drivers: HW Info>: 1, <Area Drivers: Reset>: 1, <Area West project: hal_stm32>: 1, <Area Drivers: GPIO>: 0, <Area Power management>: 0}
labels: {'area: GPIO', 'area: Power Management', 'area: Flash', 'area: HWINFO', 'area: Interrupt Controller', 'area: Clock control', 'platform: STM32'}
collab: ['erwango', 'FRASTM', 'gautierg-st', 'GeorgeCGV', 'marwaiehm-st', 'mathieuchopstm', 'nordic-krch', 'de-nordic', 'alexanderwachter', 'decsny', 'henrikbrixandersen', 'mnkp', 'ceolin', 'nashif', 'teburd', 'tmleman', 'JordanYates']
Submitted by: mathieuchopstm
candidate maintainers: {'erwango': 50, 'nordic-krch': 3, 'de-nordic': 1, 'alexanderwachter': 1, 'ceolin': 0}
```
Not sure why I wasn't picked up here ?

nashif commented:
> Not sure why I wasn't picked up here ?

general areas are being changed, general areas have priority over platform areas. To solve this issue, better file listing is needed for those areas, so that we you change something in Clock Control STM driver, it is seen as a platform change and not a subsystem change.

aescolar commented on boards/st/nucleo_wb09ke/doc/img/nucleo_wb09ke.jpg:
cc @kartben (who is welcome to override my comments)
These 2 images fit into the 100KB rule, but I'd ask they be made smaller. (we are adding more and more binary images to the tree for too little benefit)
Could you please crop some of the white away, and recompress them harder into .webp?

kartben replied:
Yes, what @aescolar said :)
```console
$ cwebp -resize 650 0 ~/Downloads/nucleo_wb05kz.jpg -o out.webp
Saving file 'out.webp'
File:      /Users/kartben/Downloads/nucleo_wb05kz.jpg
Dimension: 650 x 739
Output:    40372 bytes.  <----
```

Bonus points if you perform the conversion with the "original" pictures, if you have access to them, rather than the already compressed JPGs from this PR
Bonus points #2 if you have access to a version of these pictures with transparent background and use them instead (since webp support transparency) and also crop all the unnecessary white space around the boards.

Thanks!

mathieuchopstm replied:
I'll see what I can find :smile:

WRT image size, is there a policy I missed on recommended dimensions?
This is something I was wondering about when I added the WB05KZ image.

mathieuchopstm replied:
Changed to compressed WebP versions using https://tinypng.com (as suggested by @nordicjm) with unnecessary white space removed.

As I was only able to find JPGs, I left the white background in the images.
However, attempting to remove it manually seemed to *increase* the file size.

aescolar replied:
It is for me. Thank you :)

nordicjm commented on boards/st/nucleo_wb05kz/doc/img/nucleo_wb05kz.jpg:
Convert to webp then put through https://tinypng.com/

nordicjm commented on boards/st/nucleo_wb05kz/doc/index.rst:
```diff
@@ -0,0 +1,152 @@
+Supported Features
+==================
+
+The Zephyr nucleo_wb05kz board configuration supports the following hardware features:
```
put the board name as you use with west build, put \`\` around it and remove board configuration, it is board target

nordicjm commented on boards/st/nucleo_wb05kz/doc/index.rst:
```diff
@@ -0,0 +1,152 @@
+
+Nucleo WB05KZ board includes an ST-LINK-V3EC embedded debug tool interface.
+
+Applications for the ``nucleo_w05kz`` board configuration can be built and
```
as above

nordicjm commented on boards/st/nucleo_wb09ke/doc/index.rst:
```diff
@@ -0,0 +1,152 @@
+
+Supported Features
+==================
+
+The Zephyr nucleo_wb09ke board configuration supports the following hardware features:
```
as above and for all files in this PR

nordicjm commented on drivers/clock_control/Kconfig.stm32:
```diff
@@ -47,6 +47,50 @@ config CLOCK_STM32_MUX
+	  Setting this option to the default value of "0" disables runtime frequency
+	  measurements - the result of a single measure performed at boot will be
+	  treated as LSI frequency for the lifetime of the application.
+endmenu # DT_HAS_ST_STM32WB0_LSI_CLOCK_ENABLED
```
add newline gap

mathieuchopstm replied:
Done:
```suggestion

endmenu # DT_HAS_ST_STM32WB0_LSI_CLOCK_ENABLED
```

erwango commented:
@nordicjm PTAL
